src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,166 @@
+    private static final int DEFAULT_INITIAL_CAPACITY = 16;
+    private static final int NUMBER_TO_MULTIPLY = 2;
+    private static final int ZERO = 0;
+    private Node<K, V>[] initialArray;
+    private int size;
+    private int threshold = 12;
+
+    public MyHashMap() {
+        initialArray = new Node[DEFAULT_INITIAL_CAPACITY];
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (key == null) {
+            putNullKey(value);
+            return;
+        }
+        Node<K, V> node = newNode(key, value);
+        int index = Math.abs(getHash(key)) % initialArray.length;
+        int loader = threshold;
+        if (initialArray[index] == null) {
+            initialArray[index] = node;
+            size++;
+        } else {
+            Node<K, V> current = initialArray[index];
+            while (current.next != null) {
+                if (Objects.equals(current.key, key)) {
+                    current.value = value;
+                    return;
+                }
+                current = current.next;
+            }
+            if (Objects.equals(current.key, key)) {
+                current.value = value;
+            } else {
+                current.next = node;
+                size++;
+            }
+        }
+        if (size > loader) {
+            resize();
+        }
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        if (key == null) {
+            return getNullKey();
+        }
+        int index = getHash(key);
+        V value = null;
+        Node<K, V> currentNode = initialArray[index];
+        while (currentNode != null) {
+            if (currentNode.key != null && currentNode.key.equals(key)) {
+                value = currentNode.value;
+                break;
+            }
+            currentNode = currentNode.next;
+        }
+        return value;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private void resize() {
+        int newCapacity = initialArray.length * NUMBER_TO_MULTIPLY;
+        Node<K, V>[] newArray = new Node[newCapacity];
+        for (int i = 0; i < initialArray.length; i++) {
+            Node<K, V> current = initialArray[i];
+            while (current != null) {
+                Node<K, V> next = current.next;
+                int newIndex = current.hash % newCapacity;
+                current.next = newArray[newIndex];
+                newArray[newIndex] = current;
+                current = next;
+            }
+        }
+        initialArray = newArray;
+        threshold *= NUMBER_TO_MULTIPLY;
+    }
+
+    private V getNullKey() {
+        Node<K, V> currentNode = initialArray[ZERO];
+        while (currentNode != null) {
+            if (currentNode.key == null) {
+                return currentNode.value;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private void putNullKey(V value) {
```

Comment:
It seems to me, that this method duplicates the code in the `put`. Are you sure you need this special method just for `null` keys?